### PR TITLE
Create CRUD for customer 

### PR DESCRIPTION
**Pull request by famgorag**
- [ADD] CreateCustomer.java
- [ADD] CustomerController.java
- [ADD] CustomerRepository.java
- [ADD] CustomerResponse.java
- [ADD] CustomerService.java
- [UPD] Customer.java

### CODE REVIEW

File: src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        return customerService.getAll();
+    }
+
+    @GetMapping(value = "/{id}")
+    public CustomerResponse get(@PathVariable("id") int id) throws EntityNotFoundException {
+        return customerService.getById(id);
+    }
+
+    @DeleteMapping(value = "/{id}")
+    public void delete(@PathVariable("id") int id) {
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java:36]: Let's fill this method too

File: src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    }
+
+    @PostMapping()
+    public void create(@RequestBody CreateCustomer createCustomer) {
+
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java:42]: And this

File: src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @PutMapping(value = "/{id}")
+    public void update(@PathVariable("id") int id, @RequestBody CreateCustomer createCustomer) {
+
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java:47]: And this

File: src/main/java/com/world/entertainment/worldentertainment/dto/controller/CreateCustomer.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package com.world.entertainment.worldentertainment.dto.controller;
+
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import java.io.Serializable;
+
+final public class CreateCustomer implements Serializable {
+
+    private String name;
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/dto/controller/CreateCustomer.java:9]: This DTO is income. Make pls validation here

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        customers.forEach((entity) -> {
+            var customerResponse = new CustomerResponse(
+                    entity.getName(),
+                    entity.getEmail()
+            );
+            result.add(customerResponse);
+        });
+
+        return result;
+    }
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:50]: Let's add too Update/Delete/Create methods here

File: src/main/java/com/world/entertainment/worldentertainment/dto/controller/CreateCustomer.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    public String getEmail() {
+        return email;
+    }
+
+    public CreateCustomer setEmail(String email) {
+        this.email = email;
+        return this;
+    }
+
+    public Customer create() {
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/dto/controller/CreateCustomer.java:38]: DTO classes doesn't contain logic. Take out it to the service class

File: src/main/java/com/world/entertainment/worldentertainment/entity/Customer.java [hunk 7]
```diff
@@ -12,17 +12,15 @@
...
-
     @OneToMany(mappedBy = "customer")
     private Set<Entertainment> entertainments;
 
-    private boolean isActive;
+    private String name;
+    private String email;
 
+    private boolean isActive;
     private Date dateCreate;
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/entity/Customer.java:22]: Let's rename this fields as `deleteAt`, `updateAt` and `CreateAt`. We use this style in other classes.

File: src/main/java/com/world/entertainment/worldentertainment/repository/CustomerRepository.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.world.entertainment.worldentertainment.repository;
+
+import com.world.entertainment.worldentertainment.entity.Customer;
+import org.springframework.data.repository.CrudRepository;
+
+public interface CustomerRepository extends CrudRepository<Customer, Integer> {
+
+}
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/repository/CustomerRepository.java:8]: Add empty line in the end of this class. It's code style. We need add empty line after class body

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    public void create(Customer customer) {
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:54]: Let's to change input parameter here to `CustomCreate` and  make to transfer data from DTO to Entity here

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        customerRepository.save(customer);
+    }
+
+    public void update(Integer customerId, Customer customerUpd) {
+        var customer = customerRepository.findById(customerId).orElseThrow(()
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:59]: Let's to change input parameter to `UpdateCustomer` and change Integer to int.

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+                -> new EntityNotFoundException(String.format("Customer by id %d not found", customerId)));
+        customer
+                .setEmail(customerUpd.getEmail())
+                .setName(customerUpd.getName());
+        customerRepository.save(customer);
+    }
+
+    public void deleteById(Integer customerId) {
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:67]: Let's work with simple variable like int? We don't make something hard with ID. We don't need object here.

File: src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @PostMapping()
+    public void create(@RequestBody CreateCustomer createCustomer) {
+        customerService.create(createCustomer.create());
+    }
+
+    @PutMapping(value = "/{id}")
+    public void update(@PathVariable("id") int id, @RequestBody CreateCustomer createCustomer) {
+        customerService.update(id, createCustomer.create());
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/controller/v1/CustomerController.java:47]: Let's see adapter package, for transfer data from DTO to entity. I guess that better way for that

File: src/main/java/com/world/entertainment/worldentertainment/repository/CustomerRepository.java (hunk 8 above)
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/repository/CustomerRepository.java:8]: hmm, where empty line after class body? 
`
public interface CustomerRepository extends CrudRepository<Customer, Integer> {

}

`

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                    entity.getEmail()
+            );
+            result.add(customerResponse);
+        });
+
+        return result;
+    }
+
+    public void create(CreateCustomer createCustomer) {
+        var modelMapper = new ModelMapper();
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:58]: Wow, great thing. I don't know about it!

File: src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        var customer = modelMapper.map(createCustomer, Customer.class);
+        customerRepository.save(customer);
+    }
+
+    public void update(int customerId, UpdateCustomer updateCustomer) {
+        var customer = customerRepository.findById(customerId).orElseThrow(()
+                -> new EntityNotFoundException(String.format("Customer by id %d not found", customerId)));
+        customer
```
Comment (ViachViach) [inline src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:66]: May we can use modelMapper here too?

and can we take out model mapper to class field and inject it using constructor??

Comment (famgorag) [reply src/main/java/com/world/entertainment/worldentertainment/service/CustomerService.java:66]: Good idea. I will check it and tell you.

**Pull request merged**